nima/tests/integration/grpc/server/src/main/java/io/helidon/nima/tests/integration/grpc/webserver/EventService.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) 2023 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.nima.tests.integration.grpc.webserver;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import io.helidon.nima.grpc.events.Events;
+import io.helidon.nima.grpc.webserver.GrpcService;
+
+import com.google.protobuf.Descriptors;
+import com.google.protobuf.Empty;
+import io.grpc.stub.StreamObserver;
+
+import static io.helidon.nima.grpc.webserver.ResponseHelper.complete;
+
+public class EventService implements GrpcService {
+
+    private final List<Listener> listeners = new ArrayList<>();
+
+    private final Lock lock = new ReentrantLock();
+
+    @Override
+    public Descriptors.FileDescriptor proto() {
+        return Events.getDescriptor();
+    }
+
+    @Override
+    public void update(Routing router) {
+        router.unary("Send", this::send)
+                .bidi("Events", this::events);
+    }
+
+    private StreamObserver<Events.EventRequest> events(StreamObserver<Events.EventResponse> responses) {
+        lock.lock();
+        try {
+            Listener listener = new Listener(responses);
+            listeners.add(listener);
+            return listener;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    private void send(Events.Message message, StreamObserver<Empty> observer) {
+        String text = message.getText();
+        for (Listener listener : listeners) {
+            listener.send(text);
+        }
```

Comment:
Can `ConcurrentModificationException` happen here? Would using iterator of defensive copy make sense here?

Reply:
I added an iterator here. Also I did the same in another method of that class, and generally tidied it up and added comments to describe what it is doing.